client-sdk/ts-web/rt/src/callformat.ts

```diff
@@ -1,4 +1,93 @@
+import * as oasis from '@oasisprotocol/client';
+
+import * as transaction from './transaction';
+import * as mrae from './mrae';
+import * as types from './types';
+
 /**
  * Call data key pair ID domain separation context base.
  */
 export const CALL_DATA_KEY_PAIR_ID_CONTEXT_BASE = 'oasis-runtime-sdk/private: tx';
+
+/**
+ *  EncodeConfig is call encoding configuration.
+ *  golang: oasis-sdk/client-sdk/go/callformat/callformat.go
+ *  rust:
+ */
+export interface EncodeConfig {
+    /**
+     * publicKey is an optional runtime's call data public key to use for encrypted call formats.
+     */
+    publicKey?: types.KeyManagerSignedPublicKey;
+}
+
+export interface MetaEncryptedX25519DeoxysII {
+    sk: Uint8Array;
+    pk: Uint8Array;
+}
+
+/**
+ * encodeCall encodes a call based on its configured call format.
+ * It returns the encoded call and any metadata needed to successfully decode the result.
+ */
+export async function encodeCall(
+    nonce: Uint8Array,
+    sk: Uint8Array,
+    pk: Uint8Array,
+    call: types.Call,
+    format: types.CallFormat,
+    config?: EncodeConfig,
+): Promise<[types.Call, unknown]> {
+    switch (format) {
+        case transaction.CALLFORMAT_PLAIN:
+            return [call, undefined];
+        case transaction.CALLFORMAT_ENCRYPTED_X25519DEOXYSII:
+            if (config?.publicKey === undefined) {
+                throw new Error('callformat: runtime call data public key not set');
+            }
+            const rawCall = oasis.misc.toCBOR(call);
+            const sealedCall = mrae.boxSeal(nonce, rawCall, null, config.publicKey.key, sk);
+            const envelope: types.CallEnvelopeX25519DeoxysII = {
+                pk: pk,
+                nonce: nonce,
+                data: sealedCall,
+            };
+            const encoded: types.Call = {
+                format: transaction.CALLFORMAT_ENCRYPTED_X25519DEOXYSII,
+                method: '',
+                body: oasis.misc.toCBOR(envelope),
+            };
+            const meta: MetaEncryptedX25519DeoxysII = {
+                sk: sk,
+                pk: config.publicKey.key,
+            };
+            return [encoded, meta];
+        default:
+            throw new Error(`callformat: unsupported call format: ${format}`);
+    }
+}
+
+/**
+ * decodeResult performs result decoding based on the specified call format metadata.
+ */
+export function decodeResult(
+    result: types.CallResult,
+    meta?: MetaEncryptedX25519DeoxysII,
```

Comment:
did I leave this comment earlier? sorry if this is a duplicate.

so following on from the discussion of supporting other encodings, it may not be possible to tell what encoding was used just from this kind of metadata. for example, if there's a different encryption algorithm that also uses a public and private key in the metadata, you wouldn't be able to tell which algorithm to use.

the Go side uses different structs (instead of interfaces) and the runtime type information, which we don't have in typescript.

maybe we could have decodeResult also take a `format: types.CallFormat` parameter. would that be appropriate for your use cases? I'm imagining users will call `encodeCall`, send the tx, then call `decodeResult` in pretty close proximity, and they'll have the desired CallFormat handy in both places

Reply:
format param added, thanks